### PR TITLE
Implement BleCentralService

### PR DESCRIPTION
**Pull request by ychernysheva**
Create viable Java adapter to establish BLE connection with Android SDL (not with native code, only with MainActivity to check functionality)

Todo:

- [x] Create new Java class Service (BleCentralService).

- [x] MainActivity should start this service on click on StartSDL button via intent. 

- [x] MainActivity should stop this service on click on StopSDL button.

BleCentralService should wrap all BLE stuff, provided by Android API:
- [x] Scan
- [x] Connect
- [x] Discover services
- [x] Read characteristics
- [x] Write characteristics
- [x] Process notifications

### CODE REVIEW

File: app/src/main/java/org/luxoft/sdl_core/BleSupport/BleCentralService.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        public static final String TAG = BleCentralService.class.getSimpleName();
+
+        @Override
+        public void onCreate() {
+            super.onCreate();
+        }
+
+        @Override
+        public void onDestroy() {
+            //TODO: perform disconnect before calling onDestroy
```
Comment (AKalinich-Luxoft) [inline app/src/main/java/org/luxoft/sdl_core/BleSupport/BleCentralService.java]: @ychernysheva please check if BleCentralService throws the memory leak exception after calling onDestroy.
If it throws, then you should create a separate function that can be called before onDestroy
For better-debugging experience, please comment startSDL and stopSDL functions in the main activity so only BLE transport can be checked

File: app/src/main/java/org/luxoft/sdl_core/BleSupport/BluetoothHandler.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            instance = new BluetoothHandler(context);
+        }
+        return instance;
+    }
+
+    private final BluetoothPeripheralCallback peripheralCallback = new BluetoothPeripheralCallback() {
+        @Override
+        public void onServicesDiscovered(BluetoothPeripheral peripheral) {
+            // Request a higher MTU, iOS always asks for 185
+            peripheral.requestMtu(185);
```
Comment (AKalinich-Luxoft) [inline app/src/main/java/org/luxoft/sdl_core/BleSupport/BluetoothHandler.java]: @ychernysheva please move to private static final variable in the beginning of class

File: app/src/main/java/org/luxoft/sdl_core/BleSupport/BluetoothHandler.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                    String msg = characteristic.getStringValue(0);
+                    Log.d(BleCentralService.TAG, "message: " + msg);
+                    Handler handler = new Handler();
+                    final BluetoothPeripheral peripheral_copy = peripheral;
+                    final String msg_copy = msg;
+                    handler.postDelayed(new Runnable() {
+                        public void run() {
+                            writeMessage(peripheral_copy, "Response to: " + msg_copy);
+                        }
+                }, 3000);
```
Comment (AKalinich-Luxoft) [inline app/src/main/java/org/luxoft/sdl_core/BleSupport/BluetoothHandler.java]: @ychernysheva please add TODO comment to not forget to remove this synthetic delay in the future

File: app/src/main/java/org/luxoft/sdl_core/MainActivity.java [hunk 4]
```diff
@@ -63,13 +68,21 @@ public void onClick(View view) {
...
+                }
             }
         });
 
         stop_sdl_button.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View view) {
                 StopSDL();
+
+                if (isBleSupported()) {
```
Comment (AKalinich-Luxoft) [inline app/src/main/java/org/luxoft/sdl_core/MainActivity.java]: @ychernysheva I think there should be the same condition as for startService

File: app/src/main/java/org/luxoft/sdl_core/MainActivity.java [hunk 5]
```diff
@@ -63,13 +68,21 @@ public void onClick(View view) {
                     start_sdl_button.setEnabled(false);
                     stop_sdl_button.setEnabled(true);
                 }
+
+                if (isBleSupported() && isBluetoothPermissionGranted()) {
+                    startService(new Intent(MainActivity.this, org.luxoft.sdl_core.BleCentralService.class));
```
Comment (AKalinich-Luxoft) [inline app/src/main/java/org/luxoft/sdl_core/MainActivity.java]: @ychernysheva I think `org.luxoft.sdl_core.` prefix is redundant as these classes are in the same namespace

File: app/src/main/java/org/luxoft/sdl_core/MainActivity.java [hunk 6]
```diff
@@ -86,6 +99,10 @@ public void onClick(View view) {
                 askForStoragePermissions();
             }
         }
+
+        if (savedInstanceState == null) {
```
Comment (AKalinich-Luxoft) [inline app/src/main/java/org/luxoft/sdl_core/MainActivity.java:106]: @ychernysheva what is `savedInstanceState` and why are you checking it?

**Review by AKalinich-Luxoft — COMMENTED** [review]
Please check for the possibility to use a submodule instead of copying files from blessed repo

File: app/src/main/java/org/luxoft/sdl_core/MainActivity.java (hunk 5 above)
Comment (ychernysheva) [reply app/src/main/java/org/luxoft/sdl_core/MainActivity.java]: Done in 49c191b0

File: app/src/main/java/org/luxoft/sdl_core/MainActivity.java (hunk 6 above)
Comment (ychernysheva) [reply app/src/main/java/org/luxoft/sdl_core/MainActivity.java:106]: As I remember, variable `savedInstanceState` was created in OnCreate method of MainActivity, and in this particular case this check ensures that we are checking or asking for BT permissions only during first SDL run. But i'm unsure for that, to be honest.

File: app/src/main/java/org/luxoft/sdl_core/BleSupport/BluetoothHandler.java (hunk 2 above)
Comment (ychernysheva) [reply app/src/main/java/org/luxoft/sdl_core/BleSupport/BluetoothHandler.java]: Done in e8ca88e58483ac32df3c23ffc3c31eadcd4cab87

File: app/src/main/java/org/luxoft/sdl_core/BleSupport/BluetoothHandler.java (hunk 3 above)
Comment (ychernysheva) [reply app/src/main/java/org/luxoft/sdl_core/BleSupport/BluetoothHandler.java]: Done in e8ca88e58483ac32df3c23ffc3c31eadcd4cab87

File: app/src/main/java/org/luxoft/sdl_core/MainActivity.java (hunk 4 above)
Comment (ychernysheva) [reply app/src/main/java/org/luxoft/sdl_core/MainActivity.java]: Done in e8ca88e58483ac32df3c23ffc3c31eadcd4cab87

File: app/src/main/java/org/luxoft/sdl_core/BleSupport/BleCentralService.java (hunk 1 above)
Comment (ychernysheva) [reply app/src/main/java/org/luxoft/sdl_core/BleSupport/BleCentralService.java]: Done in 70f115a013c2b887c289081e6554d36713216229

File: app/src/main/java/org/luxoft/sdl_core/MainActivity.java [hunk 7]
```diff
@@ -41,6 +44,10 @@
     private Button stop_sdl_button;
     public static String sdl_cache_folder_path;
     public static String sdl_external_dir_folder_path;
+    private static final int ACCESS_LOCATION_REQUEST = 1;
+    private static final int ACCESS_EXT_STORAGE_REQUEST = 2;
+    public final static String ACTION_START_BLE = "ACTION_START_BLE";
+    public final static String ACTION_STOP_BLE = "ACTION_STOP_BLE";
```
Comment (AKalinich-Luxoft) [inline app/src/main/java/org/luxoft/sdl_core/MainActivity.java]: @ychernysheva actions related to service should be declared in the corresponding service. Please move these two constants to `BleCentralService`

Comment (ychernysheva) [reply app/src/main/java/org/luxoft/sdl_core/MainActivity.java]: Fixed in be07980

**Review by AKalinich-Luxoft — APPROVED** [review]

**Pull request merged**